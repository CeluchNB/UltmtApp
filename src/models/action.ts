import { DisplayUser } from '../types/user'
import { Realm } from '@realm/react'
import { TeamNumber } from '../types/team'
import { ActionType, Comment, LiveServerActionData } from '../types/action'

export const CommentSchema: Realm.ObjectSchema = {
    name: 'Comment',
    embedded: true,
    properties: {
        user: 'DisplayUser',
        comment: 'string',
        commentNumber: 'int',
    },
}

export class ActionSchema {
    static schema: Realm.ObjectSchema = {
        name: 'Action',
        primaryKey: '_id',
        properties: {
            _id: 'objectId',
            pointId: 'string',
            actionType: 'string',
            actionNumber: 'int',
            playerOne: 'DisplayUser?',
            playerTwo: 'DisplayUser?',
            tags: 'string[]',
            teamNumber: 'string',
            comments: 'Comment[]',
        },
    }

    _id: Realm.BSON.ObjectId
    pointId: string
    actionType: ActionType
    playerOne?: DisplayUser
    playerTwo?: DisplayUser
    tags: string[]
    actionNumber: number
    teamNumber: TeamNumber
    comments: Comment[]

    constructor(action: LiveServerActionData, pointId: string) {
        this._id = new Realm.BSON.ObjectID()
        this.pointId = pointId
        this.actionType = action.actionType
        this.playerOne = action.playerOne
        this.playerTwo = action.playerTwo
        this.tags = action.tags
        this.actionNumber = action.actionNumber
        this.teamNumber = action.teamNumber
        this.comments = action.comments
    }
}
