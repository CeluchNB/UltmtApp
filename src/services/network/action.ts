import { API_KEY, API_URL_V1 } from 'react-native-dotenv'
import axios, { AxiosResponse } from 'axios'

export const addComment = async (
    token: string,
    actionId: string,
    comment: string,
): Promise<AxiosResponse> => {
    return await axios.post(
        `${API_URL_V1}/action/${actionId}/comment`,
        { comment },
        {
            headers: {
                'X-API-Key': API_KEY,
                Authorization: `Bearer ${token}`,
            },
        },
    )
}
