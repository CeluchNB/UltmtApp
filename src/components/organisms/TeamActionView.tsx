import { Button } from 'react-native-paper'
import { DisplayUser } from '../../types/user'
import PlayerActionTagModal from '../molecules/PlayerActionTagModal'
import React from 'react'
import SubstitutionModal from '../molecules/SubstitutionModal'
import { useTheme } from '../../hooks'
import { Action, ActionType } from '../../types/action'
import { FlatList, StyleSheet, View } from 'react-native'

interface TeamActionViewProps {
    actions: Action[]
    onAction: (action: Action) => Promise<void>
}

const TeamActionView: React.FC<TeamActionViewProps> = ({
    actions,
    onAction,
}) => {
    const {
        theme: { colors, size },
    } = useTheme()
    const [subModalVisible, setSubModalVisible] = React.useState(false)
    const [tagModalVisible, setTagModalVisible] = React.useState(false)
    const [selectedAction, setSelectedAction] = React.useState<
        Action | undefined
    >(undefined)

    const handleAction = (action: Action) => {
        if (action.action.actionType === ActionType.SUBSTITUTION) {
            setSelectedAction(action)
            setSubModalVisible(true)
        } else {
            onAction(action)
        }
    }

    const handleLongPress = (action: Action) => {
        if (action.action.actionType === ActionType.SUBSTITUTION) {
            setSubModalVisible(true)
        } else {
            setTagModalVisible(true)
        }
        setSelectedAction(action)
    }

    const onSubModalClose = () => {
        setSubModalVisible(false)
    }

    const handleSubstitution = (
        playerOne: DisplayUser,
        playerTwo: DisplayUser,
    ) => {
        onSubModalClose()
        if (selectedAction) {
            selectedAction.setPlayers(playerOne, playerTwo)
            onAction(selectedAction)
        }
    }

    const onTagModalClose = (submit: boolean, tags: string[]) => {
        if (submit && selectedAction) {
            selectedAction.setTags(tags)
            onAction(selectedAction)
        }
        setTagModalVisible(false)
        setSelectedAction(undefined)
    }

    const styles = StyleSheet.create({
        container: {
            marginTop: 10,
            marginBottom: 10,
        },
        buttonContainer: {
            flex: 1,
        },
        button: {
            borderColor: colors.textPrimary,
            color: colors.textPrimary,
            backgroundColor: colors.primary,
            flex: 1,
            marginRight: 5,
            marginTop: 5,
        },
        buttonText: {
            fontSize: size.fontFifteen,
        },
    })

    return (
        <View style={styles.container}>
            <FlatList
                listKey="team-action-list"
                data={actions}
                numColumns={2}
                renderItem={({ item: action }) => {
                    return (
                        <View
                            key={action.action.actionType}
                            style={styles.buttonContainer}>
                            <Button
                                compact={true}
                                style={styles.button}
                                labelStyle={styles.buttonText}
                                textColor={colors.textPrimary}
                                uppercase={true}
                                collapsable={true}
                                mode="outlined"
                                onPress={() => {
                                    handleAction(action)
                                }}
                                onLongPress={() => {
                                    handleLongPress(action)
                                }}>
                                {action.reporterDisplay}
                            </Button>
                        </View>
                    )
                }}
            />
            <SubstitutionModal
                visible={subModalVisible}
                onClose={onSubModalClose}
                onSubmit={handleSubstitution}
            />
            <PlayerActionTagModal
                visible={tagModalVisible}
                onClose={onTagModalClose}
            />
        </View>
    )
}

export default TeamActionView
