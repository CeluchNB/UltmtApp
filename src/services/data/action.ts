import * as Constants from '../../utils/constants'
import EncryptedStorage from 'react-native-encrypted-storage'
import { saveActions as localSaveActions } from '../local/action'
import { addComment as networkAddComment } from '../network/action'
import { throwApiError } from '../../utils/service-utils'
import { withToken } from './auth'
import {
    ClientAction,
    SavedServerAction,
    SubscriptionObject,
} from '../../types/action'
import {
    addComment as networkAddLiveComment,
    createAction as networkCreateAction,
    joinPoint as networkJoinPoint,
    nextPoint as networkNextPoint,
    subscribe as networkSubscribe,
    undoAction as networkUndoAction,
    unsubscribe as networkUnsubscribe,
} from '../network/live-action'

/**
 * Method to join a point
 * @param gameId id of game to join
 * @param pointId id of point to join
 */
export const joinPoint = async (gameId: string, pointId: string) => {
    await networkJoinPoint(gameId, pointId)
}

/**
 * Method to create an action
 * @param action action data
 * @param pointId point action belongs to
 */
export const addAction = async (action: ClientAction, pointId: string) => {
    await networkCreateAction(action, pointId)
}

/**
 * Method to undo an action
 * @param pointId undo last action on point of this id
 */
export const undoAction = async (pointId: string) => {
    await networkUndoAction(pointId)
}

/**
 * Method to add a comment to a live action
 * @param gameId id of game
 * @param pointId id of point
 * @param actionNumber action number
 * @param teamNumber 'one' | 'two' - team reporting action
 * @param comment comment text
 */
export const addLiveComment = async (
    gameId: string,
    pointId: string,
    actionNumber: number,
    teamNumber: 'one' | 'two',
    comment: string,
) => {
    const token = (await EncryptedStorage.getItem('access_token')) || ''
    await networkAddLiveComment(
        token,
        gameId,
        pointId,
        actionNumber,
        teamNumber,
        comment,
    )
}

/**
 * Method to indicate the next point has been started
 * @param pointId id of point that was just finished
 */
export const nextPoint = async (pointId: string) => {
    await networkNextPoint(pointId)
}

/**
 * Method to listen to specific action events
 * @param subscriptions object containing subscription types and methods
 */
export const subscribe = async (subscriptions: SubscriptionObject) => {
    await networkSubscribe(subscriptions)
}

/**
 * Method to unsubscribe from all action websckets
 */
export const unsubscribe = () => {
    networkUnsubscribe()
}

/**
 * Method to add a comment to a saved action.
 * @param actionId id of action
 * @param pointId id of point
 * @param comment comment text
 * @returns updated action
 */
export const addComment = async (
    actionId: string,
    pointId: string,
    comment: string,
): Promise<SavedServerAction> => {
    try {
        const response = await withToken(networkAddComment, actionId, comment)
        const { action } = response.data
        await localSaveActions(pointId, [action])
        return action
    } catch (error) {
        return throwApiError(error, Constants.COMMENT_ERROR)
    }
}
