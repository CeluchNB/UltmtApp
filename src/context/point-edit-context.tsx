import { DisplayUser } from '../types/user'
import Point from '../types/point'
import { TeamNumber } from '../types/team'
import { finishGame } from '../services/data/game'
import { isPullingNext } from '../utils/point'
import { parseClientAction } from '../utils/action'
import usePoint from '../hooks/usePoint'
import { Action, LiveServerActionData } from '../types/action'
import { DebouncedFunc, debounce } from 'lodash'
import React, { createContext, useMemo } from 'react'
import { createPoint, finishPoint } from '../services/data/point'
import {
    resetGame,
    selectGame,
    selectTeam,
    updateScore,
} from '../store/reducers/features/game/liveGameReducer'
import {
    resetPoint,
    selectPoint,
    setPoint,
} from '../store/reducers/features/point/livePointReducer'
import { useDispatch, useSelector } from 'react-redux'

interface PointEditContextProps {
    children: React.ReactNode
}

interface PointEditContextData {
    myTeamActions: LiveServerActionData[]
    activePlayers: DisplayUser[]
    team: TeamNumber
    waiting: boolean
    game?: any
    point?: Point
    error: string
    onAction: DebouncedFunc<(action: Action) => Promise<void>>
    onUndo: DebouncedFunc<() => Promise<void>>
    onFinishPoint: () => Promise<void>
    onFinishGame: () => Promise<void>
}

export const PointEditContext = createContext<PointEditContextData>({
    myTeamActions: [],
    activePlayers: [],
    team: 'one',
    waiting: false,
    game: undefined,
    point: undefined,
    error: '',
    onAction: debounce((_action: Action) => {}),
    onUndo: debounce(() => {}),
    onFinishPoint: async () => {},
    onFinishGame: async () => {},
})

const PointEditProvider = ({ children }: PointEditContextProps) => {
    const dispatch = useDispatch()
    const game = useSelector(selectGame)
    const team = useSelector(selectTeam)
    const point = useSelector(selectPoint)
    const {
        teamOneActions,
        teamTwoActions,
        waitingForActionResponse,
        onAction,
        onNextPoint,
        onUndo,
    } = usePoint(game._id, point._id)

    const myTeamActions = useMemo(() => {
        if (team === 'one') {
            return teamOneActions
        } else {
            return teamTwoActions
        }
    }, [teamOneActions, teamTwoActions, team])

    const activePlayers = React.useMemo(() => {
        if (team === 'one') {
            return point.teamOneActivePlayers ?? []
        } else {
            return point.teamTwoActivePlayers ?? []
        }
    }, [point, team])

    const handleAction = async (action: Action) => {
        const clientAction = parseClientAction({ ...action.action })
        onAction(clientAction)
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const onActionDebounced = React.useCallback(debounce(handleAction, 150), [
        point,
    ])

    // eslint-disable-next-line react-hooks/exhaustive-deps
    const onUndoDebounced = React.useCallback(debounce(onUndo, 150), [])

    const onFinishPoint = async () => {
        const prevPoint = await finishPoint(point._id)
        const { teamOneScore, teamTwoScore } = prevPoint

        const newPoint = await createPoint(
            isPullingNext(
                team,
                myTeamActions[myTeamActions.length - 1].actionType,
            ),
            point.pointNumber + 1,
        )

        dispatch(updateScore({ teamOneScore, teamTwoScore }))
        dispatch(setPoint(newPoint))
        // if (!offline) {
        // TODO: new socket implementation
        // await nextPoint(point._id)
        // socket?.emit('point:next', JSON.stringify({ pointId: point._id }))
        onNextPoint()
        // }
    }

    const onFinishGame = async () => {
        await finishPoint(point._id)
        // if (!offline) {
        // TODO: new socket implementation
        // await nextPoint(point._id)
        onNextPoint()
        // }
        await finishGame()
        dispatch(resetGame())
        dispatch(resetPoint())
    }

    return (
        <PointEditContext.Provider
            value={{
                myTeamActions,
                activePlayers,
                team,
                game,
                point,
                waiting: waitingForActionResponse,
                error: '',
                onFinishPoint,
                onFinishGame,
                onAction: onActionDebounced,
                onUndo: onUndoDebounced,
            }}>
            {children}
        </PointEditContext.Provider>
    )
}

export default PointEditProvider
