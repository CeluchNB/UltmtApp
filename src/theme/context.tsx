import DarkTheme from './dark'
import LightTheme from './light'
import React from 'react'
import { Theme } from '../types/theme'
import { isDarkMode, setDarkMode } from '../services/data/preferences'

interface ProvidedValue {
    theme: Theme
    toggleTheme: () => void
}

export const ThemeContext = React.createContext<ProvidedValue>({
    theme: DarkTheme,
    toggleTheme: () => {},
})

interface Props {
    children?: React.ReactNode
}

export const ThemeProvider = React.memo<Props>(props => {
    const [theme, setTheme] = React.useState(DarkTheme)

    React.useEffect(() => {
        isDarkMode().then(isDark => {
            if (!isDark) {
                setTheme(LightTheme)
            }
        })
    }, [])

    const ToggleThemeCallback = React.useCallback(async () => {
        await setDarkMode(theme.id !== 'dark')
        setTheme(curr => {
            if (curr.id === 'dark') {
                console.log('setting light')
                return LightTheme
            } else {
                console.log('setting dark')
                return DarkTheme
            }
        })
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [])

    const MemoizedValue = React.useMemo(() => {
        const value: ProvidedValue = {
            theme: theme,
            toggleTheme: ToggleThemeCallback,
        }
        return value
    }, [theme, ToggleThemeCallback])

    return (
        <ThemeContext.Provider value={MemoizedValue}>
            {props.children}
        </ThemeContext.Provider>
    )
})
