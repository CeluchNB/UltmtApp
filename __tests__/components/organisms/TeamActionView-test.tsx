import { PointEditContext } from '../../../src/context/point-edit-context'
import React from 'react'
import TeamActionView from '../../../src/components/organisms/TeamActionView'
import { Action, ActionType } from '../../../src/types/action'
import {
    LiveGameContext,
    LiveGameContextData,
} from '../../../src/context/live-game-context'
import {
    fireEvent,
    render,
    screen,
    waitFor,
} from '@testing-library/react-native'

jest.mock('react-native/Libraries/Animated/NativeAnimatedHelper')

const originalWarn = console.warn.bind(console.warn)
beforeAll(() => {
    console.warn = msg =>
        !msg.toString().includes('flexWrap: `wrap`') && originalWarn(msg)
})
afterAll(() => {
    console.warn = originalWarn
})

const playerList1 = [
    {
        _id: 'realid1',
        firstName: 'First 1',
        lastName: 'Last 1',
        username: 'firstlast1',
    },
    {
        _id: 'realid2',
        firstName: 'First 2',
        lastName: 'Last 2',
        username: 'firstlast2',
    },
    {
        _id: 'realid3',
        firstName: 'First 3',
        lastName: 'Last 3',
        username: 'firstlast3',
    },
    {
        _id: 'realid4',
        firstName: 'First 4',
        lastName: 'Last 4',
        username: 'firstlast4',
    },
    {
        _id: 'realid5',
        firstName: 'First 5',
        lastName: 'Last 5',
        username: 'firstlast5',
    },
    {
        _id: 'realid6',
        firstName: 'First 6',
        lastName: 'Last 6',
        username: 'firstlast6',
    },
    {
        _id: 'realid7',
        firstName: 'First 7',
        lastName: 'Last 7',
        username: 'firstlast7',
    },
    {
        _id: 'realid8',
        firstName: 'First 8',
        lastName: 'Last 8',
        username: 'firstlast8',
    },
    {
        _id: 'realid9',
        firstName: 'First 9',
        lastName: 'Last 9',
        username: 'firstlast9',
    },
    {
        _id: 'realid10',
        firstName: 'First 10',
        lastName: 'Last 10',
        username: 'firstlast10',
    },
    {
        _id: 'realid11',
        firstName: 'First 11',
        lastName: 'Last 11',
        username: 'firstlast11',
    },
]

const actions: Action[] = [
    {
        action: {
            actionNumber: Infinity,
            actionType: ActionType.TIMEOUT,
            playerOne: playerList1[0],
            tags: [],
            comments: [],
        },
        reporterDisplay: ActionType.TIMEOUT,
        viewerDisplay: 'Timeout called',
        setTags: jest.fn(),
        setPlayersAndUpdateViewerDisplay: jest.fn(),
    },
    {
        action: {
            actionNumber: Infinity,
            actionType: ActionType.TEAM_ONE_SCORE,
            tags: [],
            comments: [],
        },
        reporterDisplay: 'they score',
        viewerDisplay: 'They score',
        setTags: jest.fn(),
        setPlayersAndUpdateViewerDisplay: jest.fn(),
    },
    {
        action: {
            actionNumber: Infinity,
            actionType: ActionType.SUBSTITUTION,
            tags: [],
            comments: [],
        },
        reporterDisplay: ActionType.SUBSTITUTION,
        viewerDisplay: 'Substitution',
        setTags: jest.fn(),
        setPlayersAndUpdateViewerDisplay: jest.fn(),
    },
]

const teamOneActivePlayers = playerList1.slice(0, 7)

const props = {
    actions,
}

beforeAll(() => {
    jest.useFakeTimers({ legacyFakeTimers: true })
})

afterAll(() => {
    jest.useRealTimers()
})

it('should match snapshot', () => {
    const snapshot = render(<TeamActionView {...props} />)
    expect(snapshot.toJSON()).toMatchSnapshot()
})

it('should handle non-substitution press', () => {
    const onAction = jest.fn()
    render(
        <PointEditContext.Provider
            value={
                {
                    onAction,
                    activePlayers: teamOneActivePlayers,
                } as any
            }>
            <TeamActionView {...props} />
        </PointEditContext.Provider>,
    )
    fireEvent.press(screen.getByText(ActionType.TIMEOUT.toString()))

    expect(onAction).toHaveBeenCalledWith(actions[0])
})

it('should handle non-substitution long press', async () => {
    const onAction = jest.fn()
    render(
        <LiveGameContext.Provider
            value={
                {
                    tags: ['huck'],
                    addTag: jest.fn(),
                } as unknown as LiveGameContextData
            }>
            <PointEditContext.Provider
                value={
                    {
                        onAction,
                        activePlayers: teamOneActivePlayers,
                    } as any
                }>
                <TeamActionView {...props} />
            </PointEditContext.Provider>
        </LiveGameContext.Provider>,
    )

    expect(screen.queryByText('Tags')).not.toBeTruthy()
    fireEvent(screen.getByText(ActionType.TIMEOUT.toString()), 'onLongPress')
    expect(screen.getByText('Tags')).toBeTruthy()

    fireEvent.press(screen.getByText('huck'))
    fireEvent.press(screen.getByText('done'))
    await waitFor(() => {
        expect(screen.queryByText('Tags')).not.toBeTruthy()
    })
    expect(actions[0].setTags).toHaveBeenCalledWith(['huck'])
    expect(onAction).toHaveBeenCalledWith(actions[0])
})

xit('should handle tag modal non submit', async () => {
    const onAction = jest.fn()
    render(
        <PointEditContext.Provider
            value={
                {
                    onAction: onAction,
                    activePlayers: teamOneActivePlayers,
                } as any
            }>
            <TeamActionView {...props} />
        </PointEditContext.Provider>,
    )

    expect(screen.queryByText('Tags')).not.toBeTruthy()
    fireEvent(screen.getByText(ActionType.TIMEOUT.toString()), 'onLongPress')
    expect(screen.getByText('Tags')).toBeTruthy()

    fireEvent(screen.getAllByTestId('base-modal')[1], 'onRequestClose')
    await waitFor(() => {
        expect(screen.queryByText('Tags')).not.toBeTruthy()
    })
    expect(onAction).not.toHaveBeenCalledWith(actions[0])
})

it('should handle substitution press', async () => {
    const onAction = jest.fn()
    render(
        <LiveGameContext.Provider
            value={{ players: playerList1 } as LiveGameContextData}>
            <PointEditContext.Provider
                value={
                    {
                        onAction: onAction,
                        activePlayers: teamOneActivePlayers,
                    } as any
                }>
                <TeamActionView {...props} />
            </PointEditContext.Provider>
        </LiveGameContext.Provider>,
    )

    expect(screen.queryByText('Player to Remove')).not.toBeTruthy()
    fireEvent.press(screen.getByText(ActionType.SUBSTITUTION.toString()))
    expect(screen.queryByText('Player to Remove')).toBeTruthy()

    fireEvent.press(screen.getByText('First 1 Last 1'))
    fireEvent.press(screen.getByText('First 10 Last 10'))
    fireEvent.press(screen.getByText('substitute'))

    await waitFor(() => {
        expect(screen.queryByText('Player to Remove')).not.toBeTruthy()
    })
    expect(actions[2].setPlayersAndUpdateViewerDisplay).toHaveBeenCalledWith(
        expect.objectContaining(playerList1[0]),
        expect.objectContaining(playerList1[9]),
    )
    expect(onAction).toHaveBeenCalledWith(actions[2])
})

it('should handle subsitution long press', async () => {
    render(
        <PointEditContext.Provider
            value={
                {
                    activePlayers: teamOneActivePlayers,
                } as any
            }>
            <TeamActionView {...props} />
        </PointEditContext.Provider>,
    )

    expect(screen.queryByText('Player to Remove')).not.toBeTruthy()
    fireEvent(
        screen.getByText(ActionType.SUBSTITUTION.toString()),
        'onLongPress',
    )
    await waitFor(() => {
        expect(screen.getByText('Player to Remove')).toBeTruthy()
    })
})
