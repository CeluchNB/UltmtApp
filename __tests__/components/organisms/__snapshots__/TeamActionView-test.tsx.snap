// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<View
  style={
    {
      "marginBottom": 10,
      "marginTop": 10,
    }
  }
>
  <RCTScrollView
    data={
      [
        "Timeout",
        "score",
        "Substitution",
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "alignSelf": undefined,
                "bottom": undefined,
                "left": undefined,
                "position": undefined,
                "right": undefined,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
                "top": undefined,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "#303030",
                    "borderColor": "#ffd600",
                    "borderRadius": 20,
                    "borderStyle": "solid",
                    "borderWidth": 1,
                    "color": "#ffd600",
                    "flex": 1,
                    "marginRight": 5,
                    "marginTop": 5,
                    "minWidth": "auto",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      false,
                      {
                        "borderRadius": 20,
                      },
                    ]
                  }
                  testID="button"
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "rgba(28, 27, 31, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          {
                            "textAlign": "left",
                          },
                          {
                            "writingDirection": "ltr",
                          },
                          [
                            {
                              "marginHorizontal": 16,
                              "marginVertical": 9,
                              "textAlign": "center",
                            },
                            false,
                            {
                              "marginHorizontal": 24,
                              "marginVertical": 10,
                            },
                            {
                              "marginHorizontal": 8,
                            },
                            {
                              "textTransform": "uppercase",
                            },
                            {
                              "color": "#ffd600",
                              "fontFamily": "System",
                              "fontSize": 14,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                            },
                            {
                              "fontSize": 30.821917808219176,
                            },
                          ],
                        ]
                      }
                    >
                      Timeout
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
          <View
            collapsable={false}
            style={
              {
                "alignSelf": undefined,
                "bottom": undefined,
                "left": undefined,
                "position": undefined,
                "right": undefined,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
                "top": undefined,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "#303030",
                    "borderColor": "#ffd600",
                    "borderRadius": 20,
                    "borderStyle": "solid",
                    "borderWidth": 1,
                    "color": "#ffd600",
                    "flex": 1,
                    "marginRight": 5,
                    "marginTop": 5,
                    "minWidth": "auto",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      false,
                      {
                        "borderRadius": 20,
                      },
                    ]
                  }
                  testID="button"
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "rgba(28, 27, 31, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          {
                            "textAlign": "left",
                          },
                          {
                            "writingDirection": "ltr",
                          },
                          [
                            {
                              "marginHorizontal": 16,
                              "marginVertical": 9,
                              "textAlign": "center",
                            },
                            false,
                            {
                              "marginHorizontal": 24,
                              "marginVertical": 10,
                            },
                            {
                              "marginHorizontal": 8,
                            },
                            {
                              "textTransform": "uppercase",
                            },
                            {
                              "color": "#ffd600",
                              "fontFamily": "System",
                              "fontSize": 14,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                            },
                            {
                              "fontSize": 30.821917808219176,
                            },
                          ],
                        ]
                      }
                    >
                      they score
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            {
              "flexDirection": "row",
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "alignSelf": undefined,
                "bottom": undefined,
                "left": undefined,
                "position": undefined,
                "right": undefined,
                "shadowColor": "#000",
                "shadowOffset": {
                  "height": 0,
                  "width": 0,
                },
                "shadowOpacity": 0,
                "shadowRadius": 0,
                "top": undefined,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "shadowColor": "#000",
                  "shadowOffset": {
                    "height": 0,
                    "width": 0,
                  },
                  "shadowOpacity": 0,
                  "shadowRadius": 0,
                }
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "#303030",
                    "borderColor": "#ffd600",
                    "borderRadius": 20,
                    "borderStyle": "solid",
                    "borderWidth": 1,
                    "color": "#ffd600",
                    "flex": 1,
                    "marginRight": 5,
                    "marginTop": 5,
                    "minWidth": "auto",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "disabled": false,
                    }
                  }
                  accessible={true}
                  collapsable={false}
                  focusable={true}
                  onBlur={[Function]}
                  onClick={[Function]}
                  onFocus={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      false,
                      {
                        "borderRadius": 20,
                      },
                    ]
                  }
                  testID="button"
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "rgba(28, 27, 31, 1)",
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          {
                            "textAlign": "left",
                          },
                          {
                            "writingDirection": "ltr",
                          },
                          [
                            {
                              "marginHorizontal": 16,
                              "marginVertical": 9,
                              "textAlign": "center",
                            },
                            false,
                            {
                              "marginHorizontal": 24,
                              "marginVertical": 10,
                            },
                            {
                              "marginHorizontal": 8,
                            },
                            {
                              "textTransform": "uppercase",
                            },
                            {
                              "color": "#ffd600",
                              "fontFamily": "System",
                              "fontSize": 14,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                            },
                            {
                              "fontSize": 30.821917808219176,
                            },
                          ],
                        ]
                      }
                    >
                      Substitution
                    </Text>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    testID="base-modal"
    transparent={true}
    visible={false}
  />
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    testID="base-modal"
    transparent={true}
    visible={false}
  />
</View>
`;
